spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ReactorSleuth.java

```diff
@@ -74,37 +78,36 @@ private ReactorSleuth() {
 				return sub;
 			}
 
-			if (beanFactory instanceof ConfigurableApplicationContext
-					&& ((ConfigurableApplicationContext) beanFactory).isActive()) {
+			if (!springContext.isActive()) {
 				if (log.isTraceEnabled()) {
-					log.trace("Spring Context [" + beanFactory
-							+ "] already refreshed. Creating a scope "
-							+ "passing span subscriber with Reactor Context " + "["
-							+ sub.currentContext() + "] and name [" + name(sub) + "]");
+					log.trace("Spring Context [" + springContext
+							+ "] is not yet refreshed. This is unexpected. Reactor Context is ["
+							+ sub.currentContext() + "] and name is [" + name(sub) + "]");
 				}
-
-				return scopePassingSpanSubscription(beanFactory, sub);
+				assert false; // should never happen, but don't break.
+				return sub;
 			}
+
 			if (log.isTraceEnabled()) {
-				log.trace("Spring Context [" + beanFactory
-						+ "] is not yet refreshed, falling back to lazy span subscriber. Reactor Context is ["
-						+ sub.currentContext() + "] and name is [" + name(sub) + "]");
+				log.trace("Spring Context [" + springContext
+						+ "] Creating a scope passing span subscriber with Reactor Context "
+						+ "[" + sub.currentContext() + "] and name [" + name(sub) + "]");
 			}
-			return new LazySpanSubscriber<>(
-					new SpanSubscriptionProvider<>(beanFactory, sub));
+
+			return scopePassingSpanSubscription(springContext, sub);
 		});
 	}
 
 	static String name(CoreSubscriber<?> sub) {
 		return Scannable.from(sub).name();
 	}
 
-	private static Map<BeanFactory, CurrentTraceContext> CACHE = new ConcurrentHashMap<>();
+	private static Map<ConfigurableApplicationContext, CurrentTraceContext> CACHE = new ConcurrentHashMap<>();
 
 	static <T> CoreSubscriber<? super T> scopePassingSpanSubscription(
-			BeanFactory beanFactory, CoreSubscriber<? super T> sub) {
-		CurrentTraceContext currentTraceContext = CACHE.computeIfAbsent(beanFactory,
-				beanFactory1 -> beanFactory1.getBean(CurrentTraceContext.class));
+			ConfigurableApplicationContext springContext, CoreSubscriber<? super T> sub) {
+		CurrentTraceContext currentTraceContext = CACHE.computeIfAbsent(springContext,
+				springContext1 -> springContext1.getBean(CurrentTraceContext.class));
```

Comment:
`springContext1 -> springContext1.getBean(CurrentTraceContext.class)` could be extracted to the field to reduce allocations (it is created on each assembly or on each last operator)

Reply:
thanks, this is killed in the other pull request.